src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.EasyRequest.cs

```diff
@@ -213,6 +213,40 @@ private void SetRedirection()
                 EventSourceTrace("Max automatic redirections: {0}", _handler._maxAutomaticRedirections);
             }
 
+            private void SetContentLength(CURLoption lengthOption)
+            {
+                Debug.Assert(lengthOption == CURLoption.CURLOPT_POSTFIELDSIZE || lengthOption == CURLoption.CURLOPT_INFILESIZE);
+
+                if (_requestMessage.Content == null)
+                {
+                    // Tell libcurl there's no data to be sent.
+                    SetCurlOption(lengthOption, 0L);
+                    return;
+                }
+
+                long? contentLengthOpt = _requestMessage.Content.Headers.ContentLength;
+                if (contentLengthOpt != null)
+                {
+                    long contentLength = contentLengthOpt.GetValueOrDefault();
+                    if (contentLength <= int.MaxValue)
+                    {
+                        // Tell libcurl how much data we expect to send.
+                        SetCurlOption(lengthOption, contentLength);
+                        return;
+                    }
+
+                    // Similarly, tell libcurl how much data we expect to send.  However,
+                    // as the amount is larger than a 32-bit value, switch to the "_LARGE"
+                    // equivalent libcurl options.
+                    SetCurlOption(
+                        lengthOption == CURLoption.CURLOPT_INFILESIZE ? CURLoption.CURLOPT_INFILESIZE_LARGE : CURLoption.CURLOPT_POSTFIELDSIZE_LARGE,
+                        contentLength);
+                    return;
```

Comment:
> is there a reason why we can't just use the *_LARGE variants exclusively?

I don't know.  I considered that, but all of the libcurl docs basically say you should use the Xx properties unless the value is too big, in which case the Xx_LARGE ones should be used.  So I figured I'd go with it.  It's possible just always using Xx_LARGE would be fine.